prefect_gcp/bigquery.py

```diff
@@ -513,3 +517,196 @@ def example_bigquery_load_file_flow():
+        else:
+            with self.gcp_credentials.get_bigquery_client() as client:
+                connection = Connection(client=client)
+                yield connection
+                connection.close()
+
+    @sync_compatible
+    async def fetch_one(
+        self,
+        operation: str,
+        parameters: Optional[Dict[str, Any]] = None,
+        cursor: Optional[Cursor] = None,
+        **execution_options: Dict[str, Any],
+    ) -> Row:
+        """
+        Fetch a single result from the database.
+
+        Args:
+            operation: The SQL query or other operation to be executed.
+            parameters: The parameters for the operation.
+            cursor: A cursor to use for the operation; if not provided, will
+                automatically create one.
+            **execution_options: Additional options to pass to `connection.execute`.
+
+        Returns:
+            A tuple containing the data returned by the database,
+                where each row is a tuple and each column is a value in the tuple.
+        """
+        with self.manage_connection() as connection:
+            cursor = cursor or connection.cursor()
+            if cursor._query_data is None:
+                await run_sync_in_worker_thread(
+                    cursor.execute,
+                    operation,
+                    parameters=parameters,
+                    **execution_options,
+                )
+            result = cursor.fetchone()
+
+        return result
+
+    @sync_compatible
+    async def fetch_many(
+        self,
+        operation: str,
+        parameters: Optional[Dict[str, Any]] = None,
+        size: Optional[int] = None,
+        cursor: Optional[Cursor] = None,
+        **execution_options: Dict[str, Any],
+    ) -> List[Row]:
+        """
+        Fetch a limited number of results from the database.
+
+        Args:
+            operation: The SQL query or other operation to be executed.
+            parameters: The parameters for the operation.
+            size: The number of results to return; if None or 0, uses the value of
+                `fetch_size` configured on the block.
+            cursor: A cursor to use for the operation; if not provided, will
+                automatically create one.
+            **execution_options: Additional options to pass to `connection.execute`.
+
+        Returns:
+            A list of tuples containing the data returned by the database,
+                where each row is a tuple and each column is a value in the tuple.
+        """
+        with self.manage_connection() as connection:
+            cursor = cursor or connection.cursor()
+            if cursor._query_data is None:
+                await run_sync_in_worker_thread(
+                    cursor.execute,
+                    operation,
+                    parameters=parameters,
+                    **execution_options,
+                )
+            size = size or self.fetch_size
```

Comment:
The `if` here is so that it doesn't re-execute the same command, but it seems overly complicated for the user to use this; I wonder if there's a better way?

A MRVE
```
from prefect_gcp.bigquery import BigQuery
from prefect_gcp.credentials import GcpCredentials

gcp_credentials = await GcpCredentials.load("gcp-credentials")
with BigQuery(gcp_credentials=gcp_credentials) as bigquery:
    with bigquery.manage_connection() as connection:
        cursor = connection.cursor()
        for i in range(0, 2):
            result = await bigquery.fetch_many(
                "SELECT * FROM `bigquery-public-data.google_trends.top_terms` LIMIT 4",
                size=2,
                cursor=cursor,
            )
            print(result, "\n")
```

Reply:
I think this class should maintain a cursor private attribute so that user's don't need to pass in a cursor themselves. We could expose a couple of methods to update the cursor too if needed (e.g. seek, reset, etc.).